### PR TITLE
Simplify Code Contributions and Reviews

### PR DESCRIPTION
**Pull request by svenefftinge**
Hey there, 👋 
this PR adds support for Gitpod, a free automated dev environment that we have built to easy the hurdles for contributions and generally working on GitHub projects. It allows anyone to get a ready-to-code dev environment for any branch, issue and pull request with a single click. 

Screenshot 2019-06-18 at 08 24 56

You can try it on this PR: 

https://gitpod.io/#https://github.com/freeCodeCamp/freeCodeCamp/pull/36302
(takes 45 secs, due to the large repo 👀)

The workspace is configured so that three processes are started automatically in distinct terminals
 - client (gatsby site)
 - api-server
 - mongodb

It seems to work well, but it would be great if someone from the team could take a closer look and tell me if there's anything to improve. Also, I'm happy to further support this in future.

Let me know what you think.

P.S.: Gitpod unleashed its full potential when you install the [GitHub app](https://github.com/marketplace/gitpod-io). It will prebuild all branches (incl. PRs) so when you open a workspace you don't have to wait for the build. The app is also capable of adding buttons to issues. For instance, you can configure it to add a button to all `first timer welcome` issues.

-----------

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

### CODE REVIEW

**Comment by raisedadead** [comment]
Hi @svenefftinge 👋  

Thanks for the PR. 

I am the lead dev here at freeCodeCamp. Can you drop me an email at `mrugesh at freecodecamp dot org` to discuss this further async or possibly hop on a call if that's something your would prefer.

We would like to learn more about the platform before we can land this?

Thanks.

**Comment by svenefftinge** [comment]
I've updated the PR and moved the `npm run seed` to every start, making sure the db is properly initialized.

**Comment by raisedadead** [comment]
Hi @ValeraS 

As one of the most prolific of contributors and having worked on the dev tooling for a while, can you possibly give this a go and see if you like this 😊 ?

Would love your feedback. Please note that the config file, which also tells GitPod how to spin up, is good to go, but may need some additional tweaks for some edge cases.

Thanks.

File: .gitpod.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  init: >
+    cd .. &&
+    wget https://fastdl.mongodb.org/linux/mongodb-linux-x86_64-3.6.3.tgz &&
+    tar xaf mongodb-linux-x86_64-3.6.3.tgz &&
+    mkdir data
+  command: /workspace/mongodb-linux-x86_64-3.6.3/bin/mongod --dbpath /workspace/data
+
+- name: client
+  before: export HOME_LOCATION=$(gp url 8000) && export API_LOCATION=$(gp url 3000)
+  init: cp sample.env .env && npm i && cd ./client && npm run predevelop && cd ..
```
Comment (ValeraS) [inline .gitpod.yml]:
```suggestion
  init: cp sample.env .env && npm ci
```

File: .gitpod.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  command: npm run seed && cd ./client && npm run develop -- -H '0.0.0.0'
```
Comment (ValeraS) [inline .gitpod.yml]:
```suggestion
  command: cd ./client && npm run develop -- -H '0.0.0.0'
```

File: .gitpod.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+- name: api-server
+  before: export HOME_LOCATION=$(gp url 8000) && export API_LOCATION=$(gp url 3000)
+  command: gp await-port 8000 && cd ./api-server && node development-entry.js
```
Comment (ValeraS) [inline .gitpod.yml]:
```suggestion
  command: gp await-port 8000 && npm run seed && npm run develop:server
```

**Review by ValeraS — COMMENTED** [review]
@svenefftinge

Does a database persist between starts? Could we run `npm run seed` only once?

A quick question, can I make new branches, switch on other branches, pull changes from remote, rebase from the master, force push changes, make new PR?

Each developer gets 100 hours per month to work on open source projects? How is time counted?

**Comment by svenefftinge** [comment]
> Does a database persist between starts? Could we run npm run seed only once?

Yes, the disk state is persisted. However mongo is configured to write to disk [only every 60 seconds](https://docs.mongodb.com/manual/faq/storage/#how-frequently-does-wiredtiger-write-to-disk). So in order to be sure the seeding is in the DB I moved it out this morning. I'll look into making sure a snapshot is taken.

> A quick question, can I make new branches, switch on other branches, pull changes from remote, rebase from the master, force push changes, make new PR?

Yes, you can do all of that.

> Each developer gets 100 hours per month to work on open source projects? How is time counted?

It is the sum of all workspace usage time. Example: If you start three workspaces in parallel and run them for an hour each you have used 3 Gitpod hours. 
The period starts from when you first logged in and renews every month. I.e. if you started using it today it renews on the 19th of every month. It is usually enough time. In those cases where people use more than 100 hours for open-source work we are happy to add extra time.

File: .gitpod.yml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+- port: 9229 # node debug
+  onOpen: ignore
+
+tasks:
+- name: db
+  before: >
+    mkdir -p /workspace/data &&
+    (~/mongodb/bin/mongod --dbpath /workspace/data --syncdelay 2 &)
+- name: server
+  before: export HOME_LOCATION=$(gp url 8000) && export API_LOCATION=$(gp url 3000)
```
Comment (ValeraS) [inline .gitpod.yml]:
```suggestion
  before: export COOKIE_DOMAIN=gitpod.io && export HOME_LOCATION=$(gp url 8000) && export API_LOCATION=$(gp url 3000)
```

File: .gitpod.yml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  init: >
+    cp sample.env .env &&
+    npm ci &&
+    npm run seed
+  command: >
+    npm run develop:server
```
Comment (ValeraS) [inline .gitpod.yml]:
```suggestion
	npm run ensure-env &&
    npm run develop:server
```

**Review by ValeraS — CHANGES_REQUESTED** [review]
@svenefftinge add this also, please.

File: .gitpod.yml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  onOpen: ignore
+- port: 3000 # api
+  onOpen: ignore
+- port: 9229 # node debug
+  onOpen: ignore
+
+tasks:
+- name: db
+  # starting mongo in background, so it doesn't block prebuilds
+  before: >
```
Comment (ValeraS) [inline .gitpod.yml]: Theoretically, can this `before` section be completed after starting the `init` server section? If so, let's place this to the server `before` section.

Comment (svenefftinge) [reply .gitpod.yml]: yes, in theory. Will update

Comment (ValeraS) [reply .gitpod.yml]: Can we await port 27017 in server `before` section?

Comment (svenefftinge) [reply .gitpod.yml]: Yes, that's possible as well. That should go before `npm run seed`, no?
Should I then keep an extra terminal for mongo?

Comment (ValeraS) [reply .gitpod.yml]: > Yes, that's possible as well. That should go before `npm run seed`, no?

Yes, and before `npm run develop:server`

**Review by ValeraS — APPROVED** [review]

**Comment by raisedadead** [comment]
Awesome, thanks for the meticulous QA Valeriy! 

@svenefftinge 

I am going to land this as soon as I have demoed this to Quincy when I meet him next week.

Looking forward to be able to recommend our user's a painless way of getting started with contributing to freeCodeCamp. 

Thanks a ton, for prep-ing this.

**Comment by svenefftinge** [comment]
@raisedadead I have resolved the merge conflict. Anything else I can do here?

**Review by raisedadead — APPROVED** [review]

**Comment by raisedadead** [comment]
Hi @svenefftinge 

Thanks a lot for preparing this PR. This will definitely help the community in setting up and working with freeCodeCamp's stack.

Cheers.

**Pull request merged**